File: experiments-local/experiment-16/runs/agent=omp_language=csharp_model=openrouter/anthropic/claude-opus-4.8_tooling=none/rep1/src/BrazilianSoccer.Core/Format.cs
// =============================================================================
// Context: Brazilian Soccer MCP Server — text formatting.
//
// Renders query-engine results into the human-readable layouts shown in the
// spec's "Example answer format" blocks (match lists with head-to-head summary,
// team records with win rate, ranked player lists, league standings, aggregate
// stats). Kept separate from both the engine (pure data) and the MCP layer
// (transport) so the same strings can be asserted in tests.
// =============================================================================
using System.Globalization;
using System.Text;

namespace BrazilianSoccer.Core;

public static class Format
{
    private static readonly CultureInfo Inv = CultureInfo.InvariantCulture;

    public static string CompetitionLabel(Competition c) => c switch
    {
        Competition.BrasileiraoSerieA => "Brasileirão Série A",
        Competition.BrasileiraoSerieB => "Brasileirão Série B",
        Competition.BrasileiraoSerieC => "Brasileirão Série C",
        Competition.CopaDoBrasil => "Copa do Brasil",
        Competition.Libertadores => "Copa Libertadores",
        _ => "Other",
    };

    public static string MatchLine(Match m)
    {
        var date = m.Date?.ToString("yyyy-MM-dd", Inv) ?? "????-??-??";
        var score = m.HasResult ? $"{m.HomeGoal}-{m.AwayGoal}" : "vs";
        var sb = new StringBuilder();
        sb.Append(date).Append(": ")
          .Append(m.HomeTeam).Append(' ').Append(score).Append(' ').Append(m.AwayTeam)
          .Append(" (").Append(CompetitionLabel(m.Competition));
        if (!string.IsNullOrEmpty(m.Round))
            sb.Append(" Round ").Append(m.Round);
        else if (!string.IsNullOrEmpty(m.Stage))
            sb.Append(' ').Append(m.Stage);
        sb.Append(')');
        return sb.ToString();
    }

    public static string Matches(IReadOnlyList<Match> matches, string? title = null, int maxLines = 25)
    {
        if (matches.Count == 0)
            return title is null ? "No matches found." : $"{title}\nNo matches found.";

        var sb = new StringBuilder();
        if (title is not null) sb.AppendLine(title);

        int shown = Math.Min(matches.Count, maxLines);
        for (int i = 0; i < shown; i++)
            sb.Append("- ").AppendLine(MatchLine(matches[i]));
        if (matches.Count > shown)
            sb.Append("- ... (").Append(matches.Count - shown).AppendLine(" more matches in dataset)");
        return sb.ToString().TrimEnd();
    }

    public static string TeamRecord(TeamRecord r, string? heading = null)
    {
        var sb = new StringBuilder();
        sb.AppendLine(heading ?? $"{r.Team} record:");
        sb.Append("- Matches: ").Append(r.Matches).AppendLine();
        sb.Append("- Wins: ").Append(r.Wins)
          .Append(", Draws: ").Append(r.Draws)
          .Append(", Losses: ").Append(r.Losses).AppendLine();
        sb.Append("- Goals For: ").Append(r.GoalsFor)
          .Append(", Goals Against: ").Append(r.GoalsAgainst).AppendLine();
        sb.Append("- Points: ").Append(r.Points).AppendLine();
        sb.Append("- Win rate: ").Append((r.WinRate * 100).ToString("0.0", Inv)).Append('%');
        return sb.ToString();
    }

    public static string HeadToHead(HeadToHead h)
    {
        var sb = new StringBuilder();
        sb.AppendLine($"Head-to-head: {h.TeamA} vs {h.TeamB}");
        sb.Append("- Matches: ").Append(h.Matches).AppendLine();
        sb.Append("- ").Append(h.TeamA).Append(" wins: ").Append(h.TeamAWins).AppendLine();
        sb.Append("- ").Append(h.TeamB).Append(" wins: ").Append(h.TeamBWins).AppendLine();
        sb.Append("- Draws: ").Append(h.Draws).AppendLine();
        sb.Append("- Goals: ").Append(h.TeamA).Append(' ').Append(h.TeamAGoals)
          .Append(", ").Append(h.TeamB).Append(' ').Append(h.TeamBGoals);
        return sb.ToString();
    }

    public static string Players(IReadOnlyList<Player> players, string? title = null, int maxLines = 25)
    {
        if (players.Count == 0)
            return title is null ? "No players found." : $"{title}\nNo players found.";

        var sb = new StringBuilder();
        if (title is not null) sb.AppendLine(title);

        int shown = Math.Min(players.Count, maxLines);
        for (int i = 0; i < shown; i++)
        {
            var p = players[i];
            sb.Append(i + 1).Append(". ").Append(p.Name)
              .Append(" - Overall: ").Append(p.Overall?.ToString(Inv) ?? "?")
              .Append(", Position: ").Append(p.Position ?? "?")
              .Append(", Club: ").Append(p.Club ?? "?")
              .Append(", Nationality: ").Append(p.Nationality)
              .AppendLine();
        }
        if (players.Count > shown)
            sb.Append("... (").Append(players.Count - shown).AppendLine(" more)");
        return sb.ToString().TrimEnd();
    }

    public static string Standings(IReadOnlyList<StandingRow> rows, string title, int maxLines = 30)
    {
        if (rows.Count == 0)
            return $"{title}\nNo data available.";

        var sb = new StringBuilder();
        sb.AppendLine(title);
        int shown = Math.Min(rows.Count, maxLines);
        for (int i = 0; i < shown; i++)
        {
            var r = rows[i];
            var rec = r.Record;
            sb.Append(r.Position).Append(". ").Append(rec.Team)
              .Append(" - ").Append(rec.Points).Append(" pts (")
              .Append(rec.Wins).Append("W, ").Append(rec.Draws).Append("D, ").Append(rec.Losses).Append("L), GD ")
              .Append(rec.GoalDifference >= 0 ? "+" : "").Append(rec.GoalDifference);
            if (i == 0) sb.Append(" - Champion");
            sb.AppendLine();
        }
        return sb.ToString().TrimEnd();
    }

    public static string Stats(CompetitionStats s, string title)
    {
        var sb = new StringBuilder();
        sb.AppendLine(title);
        sb.Append("- Matches: ").Append(s.Matches).AppendLine();
        sb.Append("- Total goals: ").Append(s.TotalGoals).AppendLine();
        sb.Append("- Average goals per match: ").Append(s.GoalsPerMatch.ToString("0.00", Inv)).AppendLine();
        sb.Append("- Home win rate: ").Append((s.HomeWinRate * 100).ToString("0.0", Inv)).Append("%").AppendLine();
        sb.Append("- Away win rate: ").Append((s.AwayWinRate * 100).ToString("0.0", Inv)).Append("%").AppendLine();
        sb.Append("- Draw rate: ").Append((s.DrawRate * 100).ToString("0.0", Inv)).Append("%");
        return sb.ToString();
    }

    public static string PlayerProfile(Player p)
    {
        var sb = new StringBuilder();
        sb.AppendLine(p.Name);
        sb.Append("- Nationality: ").Append(p.Nationality).AppendLine();
        sb.Append("- Age: ").Append(p.Age?.ToString(Inv) ?? "?").AppendLine();
        sb.Append("- Overall: ").Append(p.Overall?.ToString(Inv) ?? "?")
          .Append(", Potential: ").Append(p.Potential?.ToString(Inv) ?? "?").AppendLine();
        sb.Append("- Position: ").Append(p.Position ?? "?")
          .Append(", Jersey: ").Append(p.JerseyNumber ?? "?").AppendLine();
        sb.Append("- Club: ").Append(p.Club ?? "?").AppendLine();
        sb.Append("- Height: ").Append(p.Height ?? "?")
          .Append(", Weight: ").Append(p.Weight ?? "?");
        return sb.ToString();
    }
}

File: experiments-local/experiment-opencode-brazil/runs/agent=opencode_language=typescript_model=openrouter/z-ai/glm-5.2_tooling=none/rep2/src/format.ts
/**
 * Plain-text formatters for query results.
 *
 * The MCP tools return these human-readable strings as their primary content
 * (matching the spec's "Example answer format" blocks). The underlying
 * structured data is also attached as JSON for programmatic consumers.
 */

import type { Match, Player, TeamRecord, HeadToHead } from './types.js';
import { averageGoalsPerMatch, homeAwayRates } from './query.js';

/** Format a single match as a one-line summary. */
export function formatMatch(m: Match): string {
  const score =
    m.homeGoal != null && m.awayGoal != null
      ? `${m.homeGoal}-${m.awayGoal}`
      : '?-?';
  const parts: string[] = [];
  if (m.date) parts.push(m.date);
  parts.push(`${m.homeTeam} ${score} ${m.awayTeam}`);
  const ctx: string[] = [m.competition];
  if (m.round != null) ctx.push(`Round ${m.round}`);
  if (m.stage) ctx.push(String(m.stage));
  if (m.season != null) ctx.push(String(m.season));
  parts.push(`(${ctx.join(', ')})`);
  return parts.join(' ');
}

/** Format a list of matches with an optional header. */
export function formatMatches(matches: Match[], header?: string): string {
  const lines: string[] = [];
  if (header) lines.push(header);
  if (matches.length === 0) {
    lines.push('No matches found.');
    return lines.join('\n');
  }
  for (const m of matches) lines.push(`- ${formatMatch(m)}`);
  lines.push(`(${matches.length} match(es) in dataset)`);
  return lines.join('\n');
}

/** Format a head-to-head summary. */
export function formatHeadToHead(h2h: HeadToHead): string {
  const lines: string[] = [];
  lines.push(`${h2h.teamA} vs ${h2h.teamB}:`);
  for (const m of h2h.matches) lines.push(`- ${formatMatch(m)}`);
  lines.push('');
  lines.push(
    `Head-to-head in dataset: ${h2h.teamA} ${h2h.teamAWins} wins, ${h2h.teamB} ${h2h.teamBWins} wins, ${h2h.draws} draws`,
  );
  return lines.join('\n');
}

/** Format a TeamRecord (statistics block). */
export function formatTeamRecord(rec: TeamRecord, label?: string): string {
  const winRate =
    rec.matches > 0 ? ((rec.wins / rec.matches) * 100).toFixed(1) : '0.0';
  const lines: string[] = [];
  if (label) lines.push(label);
  lines.push(`- Matches: ${rec.matches}`);
  lines.push(`- Wins: ${rec.wins}, Draws: ${rec.draws}, Losses: ${rec.losses}`);
  lines.push(`- Goals For: ${rec.goalsFor}, Goals Against: ${rec.goalsAgainst}`);
  lines.push(`- Points: ${rec.points}`);
  lines.push(`- Win rate: ${winRate}%`);
  return lines.join('\n');
}

/** Format a standings table. */
export function formatStandings(table: TeamRecord[], title?: string): string {
  const lines: string[] = [];
  if (title) lines.push(title);
  if (table.length === 0) {
    lines.push('No standings available.');
    return lines.join('\n');
  }
  table.forEach((r, i) => {
    const gd = r.goalsFor - r.goalsAgainst;
    const champ = i === 0 ? ' - Champion' : '';
    lines.push(
      `${i + 1}. ${r.team} - ${r.points} pts (${r.wins}W, ${r.draws}D, ${r.losses}L, GD ${gd >= 0 ? '+' : ''}${gd})${champ}`,
    );
  });
  return lines.join('\n');
}

/** Format a player list (top-N). */
export function formatPlayers(
  players: Player[],
  header?: string,
): string {
  const lines: string[] = [];
  if (header) lines.push(header);
  if (players.length === 0) {
    lines.push('No players found.');
    return lines.join('\n');
  }
  players.forEach((p, i) => {
    lines.push(
      `${i + 1}. ${p.name} - Overall: ${p.overall ?? '?'}, Position: ${p.position ?? '?'}, Club: ${p.club ?? '?'}, Nationality: ${p.nationality ?? '?'}`,
    );
  });
  return lines.join('\n');
}

/** Format aggregate statistics over a match set. */
export function formatStats(matches: Match[], header?: string): string {
  const lines: string[] = [];
  if (header) lines.push(header);
  const avg = averageGoalsPerMatch(matches);
  const rates = homeAwayRates(matches);
  lines.push(`Average goals per match: ${avg.toFixed(2)}`);
  lines.push(`Home win rate: ${(rates.homeWinRate * 100).toFixed(1)}%`);
  lines.push(`Draw rate: ${(rates.drawRate * 100).toFixed(1)}%`);
  lines.push(`Away win rate: ${(rates.awayWinRate * 100).toFixed(1)}%`);
  lines.push(`Total matches: ${rates.total}`);
  return lines.join('\n');
}
